design/design-doc-operations.md

```diff
@@ -0,0 +1,501 @@
+1. Create an Operation when a resource changes using a WatchOperation
+
+Composing an Operation is hopefully self-explanatory. Operation doesn't have an
+abstraction resource (like an XR) by design - see [Alternatives
+Considered][#alternatives-considered] for more details.
+
+CronOperation and WatchOperation will be to Operation as CronJob is to Job in
+Kubernetes. They'll contain a template for an Operation, and create one as
+needed. Either on a cron schedule, or when a watched resource changes.
+
+Here's an example CronOperation:
+
+```yaml
+apiVersion: daytwo.crossplane.io/v1alpha1
+kind: CronOperation
+metadata:
+  name: cluster-rolling-upgrade
+spec:
+  # Crontab schedule
+  schedule: "0 12 * * *"
+  # How long after the scheduled time to wait before considering it too late to
+  # create the Operation.
+  startDeadline: 10m
+  # How many completed Operations to keep around.
+  successfulHistoryLimit: 3
+  failedHistoryLimit: 3
+  # Specifies how to treat concurrent executions of an operation - Allow
+  # (default), Forbid, or Replace.
+  concurrencyPolicy: Forbid
+  operationTemplate:
+    spec:
+      # How many times the pipeline can fail before the Operation fails.
+      failureLimit: 5
+      # A pipeline of functions - just like a Composition.
+      mode: Pipeline
+      pipeline:
+      - step: rolling-upgrade
+        functionRef:
+          name: function-rolling-upgrade
+        input:
+          targets:
+            apiVersion: example.org/v1
+            kind: KubernetesCluster
+            selector:
+              matchLabels:
+                daytwo.crossplane.io/eligible-for-rolling-update: "true"
+          batches:
+          - 0.01  # First upgrade 1% of clusters
+          - 0.1   # Then if that works, 10%
+          - 0.5   # Then 50%
+          - 1.0   # Then 100%
+          fromVersions:
+          - "v1.29"
+          toVersion: "v1.30"
+          versionField: spec.version
+          healthyConditions:
+          - Synced
+          - Ready
+status:
+  # Operations that're currently running
+  active:
+  - name: cluster-rolling-upgrade-amfka
+  lastScheduleTime: "2024-04-18T12:00:37+00:00"
+  lastSuccessfulTime: "2024-04-18T12:00:37+00:00"
+```
+
+This CronOperation makes a lot more sense than a bare Operation for the
+hypothetical rolling upgrade scenario. An Operation isn't long-running - it's
+akin to a single reconcile loop. So to upgrade a fleet of clusters in four
+batches you'd want the Operation to run (at least) four times, with each
+Operation handling the next largest batch.
+
+Here's an example of a WatchOperation:
+
+```yaml
+apiVersion: daytwo.crossplane.io/v1alpha1
+kind: WatchOperation
+metadata:
+  name: backup-database-on-delete
+spec:
+  # Watch for all DatabaseInstance XRs
+  apiVersion: example.org/v1
+  kind: DatabaseInstance
+  matchLabels:
+    daytwo.crossplane.io/backup-on-delete: "true"
+  # What we're watching for changes - ResourceVersion or Generation.
+  watch: ResourceVersion
```

Comment:
I wonder if this is granular enough? It may not matter to Crossplane but for an end user I could imagine it being much easier to say "I only want to trigger an operation function when X, Y or Z field has changed"

Reply:
this is a good idea. Not to hard to implement.

Reply:
Added as a future improvement.